torchtitan/parallelisms/parallelize_llama.py

```diff
@@ -295,142 +295,191 @@ def pipeline_llama_tracer(
-            row_parallel_strategy,
-            col_parallel_strategy,
-            prepare_module_input,
-        ) = get_tp_parallel_strategy(job_config)
-        loss_parallel = parallel_dims.loss_parallel_enabled
-
-        # 1. Parallelize the first embedding and the last linear proj layer
-        # 2. Parallelize the root norm layer over the sequence dim
-        # 3. Shard the first transformer block's inputs
-        model = parallelize_module(
-            model,
-            tp_mesh,
-            {
-                "tok_embeddings": RowwiseParallel(
-                    input_layouts=Replicate(),
-                    output_layouts=Shard(1),
-                ),
-                "output": col_parallel_strategy(
-                    input_layouts=Shard(1),
-                    output_layouts=Shard(-1) if loss_parallel else Replicate(),
-                    use_local_output=not loss_parallel,
-                ),
-                "norm": SequenceParallel(),
-            },
+    if job_config.model.norm_type == "fused_rmsnorm":
+        raise NotImplementedError(
+            "fused_rmsnorm not yet compatible with TP. Please use layernorm or rmsnorm."
         )
 
-        # Apply tensor + sequence parallelism to every transformer block
-        for layer_id, transformer_block in model.layers.items():
-            layer_plan = {
-                "attention": prepare_module_input(
-                    input_layouts=(Shard(1), None),
-                    desired_input_layouts=(Replicate(), None),
-                ),
-                "attention.wq": col_parallel_strategy(),
-                "attention.wk": col_parallel_strategy(),
-                "attention.wv": col_parallel_strategy(),
-                "attention.wo": row_parallel_strategy(output_layouts=Shard(1)),
-                "attention_norm": SequenceParallel(),
-                "feed_forward": prepare_module_input(
-                    input_layouts=(Shard(1),),
-                    desired_input_layouts=(Replicate(),),
-                ),
-                "feed_forward.w1": col_parallel_strategy(),
-                "feed_forward.w2": row_parallel_strategy(output_layouts=Shard(1)),
-                "feed_forward.w3": col_parallel_strategy(),
-                "ffn_norm": SequenceParallel(),
-            }
-
-            # Adjust attention module to use the local number of heads
-            attn_layer = transformer_block.attention
-            attn_layer.n_heads = attn_layer.n_heads // tp_mesh.size()
-            attn_layer.n_kv_heads = attn_layer.n_kv_heads // tp_mesh.size()
-
-            parallelize_module(
-                module=transformer_block,
-                device_mesh=tp_mesh,
-                parallelize_plan=layer_plan,
-            )
+    tp_mesh = world_mesh["tp"]
+    (
+        row_parallel_strategy,
+        col_parallel_strategy,
+        prepare_module_input,
+    ) = get_tp_parallel_strategy(job_config)
+    loss_parallel = parallel_dims.loss_parallel_enabled
```

Comment:
no need to pass in `parallel_dims`, just pass in `parallel_dims.loss_parallel_enabled` as `loss_parallel`

Reply:
to be resolved in a bigger `parallel_context` thing